mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,306 @@
+        elif op == "max":
+            self._discr_reduction = partial(self.discr.nodal_max, "vol")
+            self.rank_aggr = max
+        elif op == "norm":
+            self._discr_reduction = partial(self.discr.norm)
+            self.rank_aggr = max
+        else:
+            raise RuntimeError(f"unknown operation {op}")
+
+    @property
+    def default_aggregator(self):
+        """Rank aggregator to use."""
+        return self.rank_aggr
+
+    def __call__(self):
+        """Return the requested quantity."""
+        if self.state_vars is None:
+            return None
+
+        quantity = self.state_vars[self.quantity]
+
+        if self.dim is not None:  # momentum
+            return self._discr_reduction(quantity[self.dim])
+        else:  # mass, energy
+            return self._discr_reduction(quantity)
+
+# }}}
+
+
+# {{{ Kernel profile quantities
+
+class KernelProfile(MultiLogQuantity):
+    """Logging support for statistics of the OpenCL kernel profiling (time, \
+    num_calls, flops, bytes_accessed, footprint).
+
+    Parameters
+    ----------
+    actx
+        The array context from which to collect statistics. Must have profiling
+        enabled in the OpenCL command queue.
+
+    kernel_name
+        Name of the kernel to profile.
+    """
+
+    def __init__(self, actx: PyOpenCLArrayContext,
+                 kernel_name: str) -> None:
+        from mirgecom.profiling import PyOpenCLProfilingArrayContext
+        assert isinstance(actx, PyOpenCLProfilingArrayContext)
+
+        units = ["s", "GFlops", "1", "GByte", "GByte"]
+        names = [f"{kernel_name}_time", f"{kernel_name}_flops",
+                 f"{kernel_name}_num_calls", f"{kernel_name}_bytes_accessed",
+                 f"{kernel_name}_footprint"]
+
+        super().__init__(names, units)
+
+        self.kernel_name = kernel_name
+        self.actx = actx
+
+    def __call__(self) -> list:
+        """Return the requested kernel profile quantity."""
+        return self.actx.get_profiling_data_for_kernel(self.kernel_name)
+
+# }}}
+
+
+# {{{ Memory profiling
+
+class PythonMemoryUsage(LogQuantity):
+    """Logging support for Python memory usage (RSS, host).
+
+    Uses :mod:`memory_profiler` to track memory usage. Virtually no overhead.
+    """
+
+    def __init__(self, name: str = None):
+
+        if name is None:
+            name = "memory_usage"
+
+        super().__init__(name, "MByte", description="Memory usage (RSS, host)")
+
+        # Make sure the memory_profiler module is available
+        import importlib
+        found = importlib.util.find_spec("memory_profiler")
+        if found is None:
+            raise ImportError("memory_profiler module not found. "
+                "Install it with 'pip install memory-profiler'.")
+
+    def __call__(self) -> float:
+        """Return the memory usage."""
+        from memory_profiler import memory_usage  # pylint: disable=import-error
+        return memory_usage(-1, 0)[0]
```

Comment:
Look at [what this does under the hood](https://github.com/pythonprofilers/memory_profiler/blob/685b5f8db2a838cd324e0d686cdf7f5290281706/memory_profiler.py#L232-L437). Do you really want all that to happen on every time step? Note the timing while loops. We can probably just use `psutil` directly: https://github.com/pythonprofilers/memory_profiler/blob/685b5f8db2a838cd324e0d686cdf7f5290281706/memory_profiler.py#L126-L143

Reply:
good point, changed to psutil.